packages/shared/src/features/payment/hooks/usePaddle.ts

```diff
@@ -0,0 +1,138 @@
+import { useState, useEffect, useRef } from 'react';
+import type { Environments, Paddle, PaddleEventData } from '@paddle/paddle-js';
+import {
+  CheckoutEventNames,
+  getPaddleInstance,
+  initializePaddle,
+} from '@paddle/paddle-js';
+import { usePixelsContext } from '../../../contexts/PixelsContext';
+import { LogEvent } from '../../../lib/log';
+import { checkIsExtension } from '../../../lib/func';
+import { usePlusSubscription } from '../../../hooks/usePlusSubscription';
+
+export type PaddleEventCallback = (event: PaddleEventData) => void;
+
+function getPaddleEventCallback(
+  logSubscriptionEvent: ReturnType<
+    typeof usePlusSubscription
+  >['logSubscriptionEvent'],
+  trackPayment: ReturnType<typeof usePixelsContext>['trackPayment'],
+  paddleCallback: PaddleEventCallback,
+): PaddleEventCallback {
+  return (event: PaddleEventData) => {
+    switch (event?.name) {
+      case CheckoutEventNames.CHECKOUT_PAYMENT_INITIATED:
+        logSubscriptionEvent({
+          event_name: LogEvent.InitiatePayment,
+          target_id: event?.data?.payment.method_details.type,
+        });
+        break;
+      case CheckoutEventNames.CHECKOUT_LOADED:
+        logSubscriptionEvent({
+          event_name: LogEvent.InitiateCheckout,
+          target_id: event?.data?.payment.method_details.type,
+        });
+        break;
+      case CheckoutEventNames.CHECKOUT_PAYMENT_SELECTED:
+        logSubscriptionEvent({
+          event_name: LogEvent.SelectCheckoutPayment,
+          target_id: event?.data?.payment.method_details.type,
+        });
+        break;
+      case CheckoutEventNames.CHECKOUT_COMPLETED:
+        logSubscriptionEvent({
+          event_name:
+            'gifter_id' in event.data.custom_data
+              ? LogEvent.CompleteGiftCheckout
+              : LogEvent.CompleteCheckout,
+          extra: {
+            user_id:
+              'gifter_id' in event.data.custom_data &&
+              'user_id' in event.data.custom_data
+                ? event.data.custom_data.user_id
+                : undefined,
+            cycle: event?.data.items?.[0]?.billing_cycle?.interval ?? 'one-off',
+            localCost: event?.data.totals.total,
+            localCurrency: event?.data.currency_code,
+            payment: event?.data.payment.method_details.type,
+          },
+        });
+        trackPayment(
+          event?.data.totals.total,
+          event?.data.currency_code,
+          event?.data?.transaction_id,
+        );
+        break;
+      // This doesn't exist in the original code
+      case 'checkout.warning' as CheckoutEventNames:
+        logSubscriptionEvent({
+          event_name: LogEvent.WarningCheckout,
+        });
+        break;
+      case CheckoutEventNames.CHECKOUT_ERROR:
+        logSubscriptionEvent({
+          event_name: LogEvent.ErrorCheckout,
+        });
+        break;
+      default:
+        break;
+    }
+    paddleCallback(event);
```

Comment:
This is pretty neat :) 
It does execute both?

Haven't ever tested 2 of the same switch statement vs early breaks.

Reply:
Awesome good to know, we can use this pattern more often :) 
The hook can do it's always needed stuff and expose extra callback switch 🤯

Reply:
Yap. i love this pattern